### PR TITLE
Update warp.py

### LINKED ISSUE #31276: Problems with Warp (Reproject) and user-defined output extent
**Description**

A problem occurs when using Warp (Reproject) from the Processing Toolbox or  gdal:warpreproject in Python when the CRS of a user-defined output extent is implicit or is specified in a user-defined coordinate system. In the former case the GDAL/OGR console call contains the ‘-te_srs’ string but no CRS definition.  In the latter case the -te_srs parameter is specified as an authid rather than being translated into a Proj4 string and this presumably causes the GDAL routine to crash. The error message in both cases is 'Translating source or target SRS failed'. If the Source CRS and Target CRS are explicitly specified as user-defined, these are translated to Proj4 in the console call.

**How to reproduce (implicit projection)**

Run Warp (Reproject) from the Processing Toolbox on any suitable raster and a define a georeferenced extent for the output in Advanced Parameters (e.g. the canvas extent) while leaving the CRS of the target raster extent blank.

**How to reproduce (user-defined projection)**

1.  Load a raster with a user-defined CRS into QGIS (example provided).
2.  Run Warp (Reproject) and select the user-defined CRS in the Source, Target and Extent drop-down lists. Define an extent for the output.
3.  Note the difference in the way the GDAL/OGR console call defines the -s_srs, -t_srs and -te_srs parameters.
4.  The routine crashes if -te_srs is a USER: rather than EPSG: system. 
[test_uklidar.zip](https://github.com/qgis/QGIS/files/3509481/test_uklidar.zip)

**QGIS and OS versions**

QGIS version
3.8.1-Zanzibar
QGIS code revision
dcd95cc648
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
2.4.1
Running against GDAL/OGR
2.4.1
Compiled against GEOS
3.7.2-CAPI-1.11.0
Running against GEOS
3.7.2-CAPI-1.11.0 b55d2125
PostgreSQL Client Version
9.2.4
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
5.2.0
Running against PROJ
Rel. 5.2.0, September 15th, 2018
OS Version
Windows 10 (10.0)

**Additional context**

Problem also occurs with the 'processing.run('gdal:warpreproject', parameters)' Python command  (TARGET_EXTENT_CRS parameter)

### PR DESCRIPTION
**Pull request by gkimbell**
## Description
Changes proposed to address bug reported in https://github.com/qgis/QGIS/issues/31276. The change to line 218 addresses the case where an output extent is defined but its CRS is left blank (implicit). The change to line 220 ensures that a USER: coordinate system is translated into a Proj4 string that is recognised by GDAL.

Apologies if this pull request doesn't comply with some items on the checklist (I'm unfamiliar with this process)

Fixes #31276

## Checklist

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by roya0045** [comment]
Have you tested this?

I assume that the best way to ensure that this work would be to make a test in there https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py

**Comment by nyalldawson** [comment]
Fix looks good!

As @roya0045 has pointed out, we also need a unit test covering this. It would slot in right near these similar tests: https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py#L1638

File: python/plugins/processing/tests/GdalAlgorithmsRasterTest.py [hunk 1]
```diff
@@ -1659,6 +1659,20 @@ def testWarp(self):
...
+                                        'SOURCE_CRS': custom_crs,
+                                        'TARGET_CRS': custom_crs,
+                                        'TARGET_EXTENT': '18.67,18.70,45.78,45.81',
+                                        'TARGET_EXTENT_CRS': custom_crs,
+                                        'OUTPUT': outdir + '/check.jpg'}, context, feedback),
+                ['gdalwarp',
+                 '-s_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -t_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -r near -te 18.67 45.78 18.7 45.81 -te_srs "+proj=utm +zone=36 +south +a=63785 +b=6357 +towgs84=-143,-90,-294,0,0,0,0 +units=m +no_defs" -of JPEG ' +
+                 source + ' ' +
+                 outdir + '/check.jpg'])
+
```
Comment (roya0045) [inline python/plugins/processing/tests/GdalAlgorithmsRasterTest.py]: I would suggest you to create a distinct test. Basically copy this function, and remove elements that belong to the other test and validate elements related to your changes. This way we can more easily monitor when something breaks this functionality.

**Comment by gkimbell** [comment]
I've added a tentative unit test at lines 1662-1674 of https://github.com/gkimbell/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsRasterTest.py#L1662 

Not too confident about this as I have been unable to run the unit test properly on my PC (combination of trying to do things in Windows and extreme incompetence). Running an equivalent test manually in QGIS Desktop works OK after the edits to warp.py.

**Comment by roya0045** [comment]
Since there are so many tests in this classes it should be fine, I didn't think that all the warp parameters were tested in the same function. This should be fine.

**Comment by roya0045** [comment]
@gkimbell could you do a minimal commit to trigger to force travis to build and test again. I don't think your changes are what is causing the test to fail. If you force travis to rebuild well be able to know if everything is ok and the devs will be able to merge it once they get the green check mark.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**